code/modules/hydroponics/seed_extractor.dm

```diff
@@ -107,107 +109,127 @@
-		for (var/datum/seed_pile/O in piles)
-			dat += "<tr><td>[O.name]</td><td>[O.variant]</td><td>[O.lifespan]</td><td>[O.endurance]</td><td>[O.maturation]</td>"
-			dat += "<td>[O.production]</td><td>[O.yield]</td><td>[O.potency]</td><td>"
-			dat += "<a href='byond://?src=[src.UID()];name=[O.name];li=[O.lifespan];en=[O.endurance];ma=[O.maturation];pr=[O.production];yi=[O.yield];pot=[O.potency]'>Vend</a> ([O.amount] left)</td></tr>"
-		dat += "</table>"
-	var/datum/browser/popup = new(user, "seed_ext", name, 700, 400)
-	popup.set_content(dat)
-	popup.open()
-	return
-
-/obj/machinery/seed_extractor/Topic(href, list/href_list)
-	if(..())
-		return 1
-	usr.set_machine(src)
-
-	href_list["li"] = text2num(href_list["li"])
-	href_list["en"] = text2num(href_list["en"])
-	href_list["ma"] = text2num(href_list["ma"])
-	href_list["pr"] = text2num(href_list["pr"])
-	href_list["yi"] = text2num(href_list["yi"])
-	href_list["pot"] = text2num(href_list["pot"])
-
-	for (var/datum/seed_pile/N in piles)//Find the pile we need to reduce...
-		if (href_list["name"] == N.name && href_list["li"] == N.lifespan && href_list["en"] == N.endurance && href_list["ma"] == N.maturation && href_list["pr"] == N.production && href_list["yi"] == N.yield && href_list["pot"] == N.potency)
+/obj/machinery/seed_extractor/proc/vendSeed(seed_id, amount)
+	if(!seed_id)
+		return
+	var/datum/seed_pile/selected_pile
+	for(var/datum/seed_pile/N in piles)
+		if(N.id == seed_id)
+			N.amount -= amount
+			selected_pile = N
 			if(N.amount <= 0)
-				return
-			N.amount = max(N.amount - 1, 0)
-			if (N.amount <= 0)
 				piles -= N
-				qdel(N)
 			break
-
-	for (var/obj/T in contents)//Now we find the seed we need to vend
+	if(!selected_pile)
+		return
+	var/amount_dispensed = 0
+	for(var/obj/T in contents)
+		if(amount_dispensed >= amount)
+			break
 		var/obj/item/seeds/O = T
-		if (O.plantname == href_list["name"] && O.lifespan == href_list["li"] && O.endurance == href_list["en"] && O.maturation == href_list["ma"] && O.production == href_list["pr"] && O.yield == href_list["yi"] && O.potency == href_list["pot"])
+		if (O.plantname == selected_pile.name && O.lifespan == selected_pile.lifespan && O.endurance == selected_pile.endurance && O.maturation == selected_pile.maturation && O.production == selected_pile.production && O.yield == selected_pile.yield && O.potency == selected_pile.potency)
 			O.forceMove(loc)
-			break
-
-	updateUsrDialog()
-	return
+			amount_dispensed++
 
 /obj/machinery/seed_extractor/proc/add_seed(obj/item/seeds/O)
-	if(contents.len >= 999)
-		to_chat(usr, "<span class='notice'>\The [src] is full.</span>")
-		return 0
-
-	if(istype(O.loc,/mob))
-		var/mob/M = O.loc
-		if(!M.drop_item())
-			return 0
-	else if(istype(O.loc,/obj/item/storage))
+	if(!O || !ishuman(usr) || !Adjacent(usr))
+		return
+	if(length(contents) >= max_seeds)
+		to_chat(usr, "<span class='notice'>[src] is full.</span>")
+		return FALSE
+
+	if(ismob(O.loc))
+		usr.drop_item()
```

Comment:
I'd rather user be passed as argument to `add_seed` instead of checking raw `usr`, but I can buy this being out of scope.

However, this particular check looks pretty scary since there's seemingly no connection between `O` and `usr`.
I'd prefer it to be like the old behavior, i.e.:
```suggestion
	if(ismob(O.loc))
		var/mob/M = O.loc
		M.drop_item()
```

But even still: Why do we need this `.drop_item()` / `.remove_from_storage()` dance in the first place? Does `O.forceMove(src)` not handle that for us via the magic of signals? (i don't know)

Reply:
im not 100% sure why we use this, I think it's just better practice? A lot of other procs in other objects use this.

Reply:
I don't think it's good practice.
If it was good practice, then every single `item.forceMove()` call would need to invoke this 6-line check or risk triggering bugs. And at that point, why not move the check to `.forceMove` and protect ourselves from forgetting to add it at callsite?

Either way, is technically out of scope.